libraries/standard/mqtt/src/iot_mqtt_api.c

```diff
@@ -1541,85 +1515,76 @@ IotMqttError_t IotMqtt_PublishAsync( IotMqttConnection_t mqttConnection,
-    {
-        goto cleanup;
+        /* Generate a PUBLISH packet from pPublishInfo. */
+        status = _getMqttPublishSerializer( mqttConnection->pSerializer )( pPublishInfo,
+                                                                           &( pOperation->u.operation.pMqttPacket ),
+                                                                           &( pOperation->u.operation.packetSize ),
+                                                                           &( pOperation->u.operation.packetIdentifier ),
+                                                                           pPacketIdentifierHigh );
     }
 
-    /* Check the serialized MQTT packet. */
-    IotMqtt_Assert( pOperation->u.operation.pMqttPacket != NULL );
-    IotMqtt_Assert( pOperation->u.operation.packetSize > 0 );
-
-    /* Initialize PUBLISH retry if retryLimit is set. */
-    if( pPublishInfo->retryLimit > 0 )
+    if( status == IOT_MQTT_SUCCESS )
     {
-        /* A QoS 0 PUBLISH may not be retried. */
-        if( pPublishInfo->qos != IOT_MQTT_QOS_0 )
+        /* Check the serialized MQTT packet. */
+        IotMqtt_Assert( pOperation->u.operation.pMqttPacket != NULL );
+        IotMqtt_Assert( pOperation->u.operation.packetSize > 0 );
+
+        /* Initialize PUBLISH retry if retryLimit is set. */
+        if( pPublishInfo->retryLimit > 0 )
         {
-            pOperation->u.operation.periodic.retry.limit = pPublishInfo->retryLimit;
-            pOperation->u.operation.periodic.retry.nextPeriodMs = pPublishInfo->retryMs;
+            /* A QoS 0 PUBLISH may not be retried. */
+            if( pPublishInfo->qos != IOT_MQTT_QOS_0 )
+            {
+                pOperation->u.operation.periodic.retry.limit = pPublishInfo->retryLimit;
+                pOperation->u.operation.periodic.retry.nextPeriodMs = pPublishInfo->retryMs;
+            }
         }
-    }
 
-    /* Set the reference, if provided. */
-    if( pPublishInfo->qos != IOT_MQTT_QOS_0 )
-    {
-        _setOperationReference( pPublishOperation, pOperation );
-    }
+        /* Set the reference, if provided. */
+        if( pPublishInfo->qos != IOT_MQTT_QOS_0 )
+        {
+            _setOperationReference( pPublishOperation, pOperation );
+        }
 
-    /* Send the PUBLISH packet. */
-    if( ( flags & MQTT_INTERNAL_FLAG_BLOCK_ON_SEND ) == MQTT_INTERNAL_FLAG_BLOCK_ON_SEND )
-    {
-        _IotMqtt_ProcessSend( IOT_SYSTEM_TASKPOOL, pOperation->job, pOperation );
-    }
-    else
-    {
-        status = _IotMqtt_ScheduleOperation( pOperation,
-                                             _IotMqtt_ProcessSend,
-                                             0 );
+        /* Send the PUBLISH packet. */
+        if( ( flags & MQTT_INTERNAL_FLAG_BLOCK_ON_SEND ) == MQTT_INTERNAL_FLAG_BLOCK_ON_SEND )
+        {
+            _IotMqtt_ProcessSend( IOT_SYSTEM_TASKPOOL, pOperation->job, pOperation );
+        }
+        else if( ( ( status = _IotMqtt_ScheduleOperation( pOperation,
```

Comment:
Is this combined assign/compare also a MISRA violation?

Reply:
Yes, it is (13.4) . Along with goto removal, we have to also make sure complexity is below 8. This is one of those functions where we have complexity score more than 8, So I opted for using assign and compare, in order to keep it less. I'll add annotation for these for now. I'll revisit when I take another shot at complexity reduction.